ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -378,30 +380,31 @@ private byte[] serializeKey(
    * schemaID, meaning that logically identical keys might be routed to
    * different partitions.
    */
-  private static void ensureKeySchemasMatch(
+  private static Optional<Integer> ensureKeySchemasMatch(
       final PersistenceSchema keySchema,
       final DataSource dataSource,
       final ServiceContext serviceContext
   ) {
     final KeyFormat keyFormat = dataSource.getKsqlTopic().getKeyFormat();
     final Format format = FormatFactory.fromName(keyFormat.getFormat());
     if (!format.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)) {
-      return;
+      return Optional.empty();
     }
 
     final SchemaRegistryClient schemaRegistryClient = serviceContext.getSchemaRegistryClient();
 
     final FormatInfo formatInfo = addSerializerMissingFormatFields(
         dataSource.getKsqlTopic().getKeyFormat().getFormatInfo(),
         dataSource.getKafkaTopicName(),
-        true
+        true,
+        Optional.empty()
     );
 
     final ParsedSchema schema = format
         .getSchemaTranslator(formatInfo.getProperties())
         .toParsedSchema(keySchema);
 
-    final Optional<ParsedSchema> latest;
+    final SchemaWithId latest;
     try {
       latest = SchemaRegistryUtil.getLatestParsedSchema(
```

Comment:
Could you add a comment about the reason we're not looking at the schema if KEY_SCHEMA_ID was used in the CREATE? I was confused for a little bit what would happen if we ensure the key match with the KEY_SCHEMA_ID but I recalled that that could still cause issues if the schema from KEY_SCHEMA_ID differs from the latest schema available.

Reply:
Added a comment